### PR TITLE
D-Bus API: support for paged list of available subscriptions; ENT-1762

### PR DESCRIPTION
**Pull request by jirihnidek**
* Small modification of D-Bus API to support listing of available
  subscriptions in pagged manner. It will be usefull for ver long
  list of available subscriptions in our cockpit plugin.
* D-Bus method GetPools can have two more options: "page" and
  "items_per_page". When options items_per_page is bigger than 0,
  then reduced list is returned. Example:

  sudo dbus-send --system --print-reply --dest=com.redhat.RHSM1 \
     /com/redhat/RHSM1/Entitlement com.redhat.RHSM1.Entitlement.GetPools \
     dict:string:string:"pool_subsets","available","items_per_page","3" \
     dict:string:string:"","" string:""

* When items_per_page is bigger is bigger than 0, then each item in
  the list will have three more attributes: page, pages and items_per_page
  with current values
* Caching of available entitlements
  - Added option to cache list of available entitlements.
  - When the system is registered again, then cache is obsoleted.
  - When filter options are changed, then cache is obsoleted too.
* Cache of available entitlements is obsoleted after seconds
  - Timeout is computed from smoothed response time
  - Lower bound of timeout is 5 seconds, upper bound is 10 seconds
    Use paginating results from candlepin
* It is possible to get paginating results from candlepin server
  for several REST API. More details here:

  https://www.candlepinproject.org/docs/candlepin/pagination.html

  - It is possible to use paginating result using D-Bus API, when
    items_per_page is not zero. It can be tested with this bash
    script with candlepin test data (it simulates scrolling of
    page with many available entitlements):

```bash
for page in $(seq 0 15)
do
        sleep 0.5  # delay between d-bus call is relativelly small
        dbus-send --system --print-reply --dest=com.redhat.RHSM1 \
        /com/redhat/RHSM1/Entitlement com.redhat.RHSM1.Entitlement.GetPools \
        dict:string:string:"pool_subsets","available","items_per_page","5","page","${page}" \
        dict:string:string:"","" string:""
done
```

  - It is possible to notice in rhsm.log that cache is not
    obsoleted by timeout, but it is obsoleted by different filter
    options
  - Option per_page used in REST API call is 4 times bigger then in
    D-Bus call. Thus he cached data are not obsoleted by next call
    with incremented page option.
* Added several unit tests

### CODE REVIEW

**Comment by jirihnidek** [comment]
@cnsnyder I have same concerns and I had the same idea. :smile: The question is how short/long should be the "short internval"? I propose to compute it from RTT of current connection to candlepin server using `short_interval = 4 * RTT`. Thoughts?

**Comment by cnsnyder** [comment]
@jirihnidek Why not use candlepin's paging? We could grab pages of one size (based on what we'd like to hold in memory in the browser) and forward that info through to candlepin. Then we could have a display window which pages through the larger pages in memory.

**Comment by jirihnidek** [comment]
> @jirihnidek Why not use candlepin's paging?

What is it?

**Comment by barnabycourt** [comment]
@jirihnidek Details on the Candlepin paging parameters for API calls at https://www.candlepinproject.org/docs/candlepin/pagination.html. As long as you specify an order and stick with it things should be ok.

**Comment by jirihnidek** [comment]
> @jirihnidek Why not use candlepin's paging? We could grab pages of one size (based on what we'd like to hold in memory in the browser) and forward that info through to candlepin. Then we could have a display window which pages through the larger pages in memory.

Interesting. I think that we can use it and it could improve user experience and it could also reduce traffic between client and server, but it makes things even more complicated. :-)

**Comment by cnsnyder** [comment]
@jirihnidek I agree that it complicates things a bit but I'm still convinced from a performance perspective it is the right way to go. Moving back to In Progress. If you disagree, please let me know in a comment and remove the "In Progress" label.

**Comment by jirihnidek** [comment]
OK, I have idea. I would like to first implement caching of all available entitlements and then I will add support for paginating results form candlepin server. Thoughts?

**Comment by jirihnidek** [comment]
@candlepin-bot retest it please

**Comment by jirihnidek** [comment]
@candlepin-bot test it please

**Comment by wottop** [comment]
Just a note: it's spelled "paged"

File: src/rhsmlib/services/entitlement.py [hunk 1]
```diff
@@ -330,6 +376,21 @@ def filter_pool_by_service_level(pool_data):
         if service_level is not None:
             available_pools = list(filter(filter_pool_by_service_level, available_pools))
 
+        # When pagination result of available pools is requested, then reduce too long list
+        if items_per_page > 0:
```
Comment (cnsnyder) [inline src/rhsmlib/services/entitlement.py]: I think we should revisit paging in general across our tooling. This is a good solution for now.
In the future I think the subscription tooling would benefit from improvements to the implementation of paging in Candlepin to ensure consistency for paging results from a given query and to reduce the memory overhead.

If we could be sure that we'll get responses that don't miss items we could make use of candlepin's paging APIs. The only other thing to consider is RTT (as you have here). Making use of Candlepin's paging could reduce performance significantly if the RTT is high. Could we implement a hybrid approach? If the RTT is past some limit, use local paging and grab the whole set of pools. If RTT is below that number, grab directly from candlepin always using candlepin's paging?

File: src/subscription_manager/cache.py [hunk 2]
```diff
@@ -825,3 +825,49 @@ def _load_data(self, open_file):
...
             # Ignore json file parse error
             pass
+
+
+class AvailableEntitlementsCache(CacheManager):
+    """
+    Cache of available entitlements
+    """
+
+    # Coefficient used for computing timout of cache
```
Comment (cnsnyder) [inline src/subscription_manager/cache.py]: timout -> timeout

File: src/subscription_manager/cache.py [hunk 3]
```diff
@@ -825,3 +825,49 @@ def _load_data(self, open_file):
...
+        self.available_entitlements = available_entitlements or {}
+
+    def to_dict(self):
+        return self.available_entitlements
+
+    def timeout(self):
+        """
+        Compute timeout of cache. Computation of timeout is based on SRT (smoothed response time)
+        of connection to candlepin server. This algorithm is inspired by retransmission timout used
+        byt TCP connection (see: RFC 793)
```
Comment (cnsnyder) [inline src/subscription_manager/cache.py]: byt -> by

**Comment by jirihnidek** [comment]
> This still does not account for caching of multiple different paged views the data (filter_options).
> 
> Please let me know how we should proceed. From my perspective, we could take this PR as is and capture future changes in a bug or card or issue, or we can continue the dialog.

You are right. I can add page to filter_options. I will add it as part of this PR.

**Pull request merged**